cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,85 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Public cuML interface
+### Motivation
+The public cuML interface is stateless for two main reasons:
+1. To ease the serialization of ML algorithm's state information (model, hyper-params, etc), enabling features such as easy pickling in the python layer.
+2. To provide proper C bindings for interfacing with languages that can't consume C++ APIs  directly.
+
+Thus, this section lays out guidelines for managing state along the API of cuML.
+
+### General guideline
+As mentioned before, functions exposed via the base cuML C++ layer must be stateless. Things that are OK to be exposed on the interface:
+1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - see [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) as a reference for C++11  POD types.
+2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`([reference](src/cuML_api.h)), instead.
+3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).
```

Comment:
Should this read `even though it can be considered as a POD`?

Reply:
Done.